File: Medium/2026-05-05-61-Rotate-List/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head or not head.next or k == 0:
            return head
        
        # Compute the length of the list
        length = 1
        tail = head
        while tail.next:
            tail = tail.next
            length += 1
        
        # Make the list circular
        tail.next = head
        
        # Find the new tail: (length - k % length - 1)th node
        k = k % length
        new_tail = head
        for _ in range(length - k - 1):
            new_tail = new_tail.next
        
        # Set the new head
        new_head = new_tail.next
        new_tail.next = None
        
        return new_head

File: Medium/2026-05-05-61-Rotate-List/solution.java
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) {
            return head;
        }

        // Calculate the length of the list
        ListNode current = head;
        int length = 1;
        while (current.next != null) {
            current = current.next;
            length++;
        }

        // Connect the tail to the head to make it circular
        current.next = head;

        // Find the new tail: (length - k % length - 1)th node
        k = k % length;
        int newTailIndex = length - k - 1;
        ListNode newTail = head;
        for (int i = 0; i < newTailIndex; i++) {
            newTail = newTail.next;
        }

        // The new head is the next of the new tail
        ListNode newHead = newTail.next;
        // Break the circular link
        newTail.next = null;

        return newHead;
    }
}